File: raspberry_pi/python/epd_gui.py
import os
import sys
import numpy as np
import epd_2inch13
from PIL import Image, ImageFont

FILL_EMPTY = 0
FILL_FULL  = 1

LINE_SOLID = 0
LINE_DOTTED = 1
FONT_SIZE_16 = 16
FONT_SIZE_24 = 24
FONT_SIZE_28 = 28
class EPD_GUI():
    def __init__(self):
        self.epd = epd_2inch13.EPD_2Inch13()
        self.epd.reset()
        self.epd.epd_init()
        self.img = [0xff for i in range(4000)]

    def set_pixel(self,x,y):
        if x>self.epd.dis_w or y>self.epd.dis_h:
            print("draw point out of the screen area!")
            return 0
        y += 6 #128-122=6
        byte_offset = int((y-1)/8+(x-1)*16)
        bit_offset = int((8-(y-1)%8)-1)
        temp = ~(1<<(bit_offset))
        self.img[byte_offset] &= temp
        return 1
    
    def clear(self):
        for i in range(len(self.img)):
            self.img[i] = 0xff
            
    def draw_point(self,x,y,width):
        if x>self.epd.dis_w or y>self.epd.dis_h:
            print("ERR:Gui_draw_point out of screen area!\r\n")
            return 0
        for xd in range(0,width):
            for yd in range(width):
                if (x+xd)>self.epd.dis_w or (y+yd)>self.epd.dis_h:
                   continue
                self.set_pixel(x+xd, y+yd)

    #Bresenham algorithm draw line
    def draw_line(self,x1, y1, x2, y2, width, line_type):
        x = x1
        y = y1
        dx = x2-x1 if x2-x1>=0 else x1-x2
        dy = y2-y1 if y2-y1>=0 else y1-y2
        p = dx-dy
        
        xdir = -1 if x1>x2 else 1
        ydir = -1 if y1>y2 else 1
        dot_cnt = 0
        while(True):
            dot_cnt += 1
            if line_type==LINE_DOTTED and dot_cnt%3==0:
                self.draw_point(x, y, width)
            elif line_type==LINE_SOLID:
                self.draw_point(x, y, width)
            if x==x2 and y==y2:
                break
            e2 = 2*p
            if e2>= -dy:
                p -= dy
                x += xdir
            if e2<=dx:
                p += dx
                y += ydir
        
    def draw_rectangle(self,x1, y1, x2, y2, width, fill):
        if x1>self.epd.dis_w or y1 > self.epd.dis_h or x2 > self.epd.dis_w or y2 > self.epd.dis_h:
            printf("ERR:line point out of screen area!\r\n")
            return 0
        if fill == FILL_FULL:
            for i in range(y1,y2):
                self.draw_line(x1, i, x2, i, width, LINE_SOLID)
        elif fill==FILL_EMPTY:
            self.draw_line(x1, y1, x2, y1, width, LINE_SOLID)
            self.draw_line(x1, y1, x1, y2, width, LINE_SOLID)
            self.draw_line(x1, y2, x2, y2, width, LINE_SOLID)
            self.draw_line(x2, y1, x2, y2, width, LINE_SOLID)
            
    #Bresenham  algorithm draw circle
    def draw_circle(self, x, y, r, width, fill):
        dx=0
        dy=r
        d = 1-r
        if x>self.epd.dis_w  or y > self.epd.dis_h:
            print("ERR:circle center point out of screen area!\r\n")
            return 0
        while(dy>dx):
            if fill==FILL_EMPTY:
                self.draw_point(x+dx, y+dy, width)
                self.draw_point(x+dy, y+dx, width)
                self.draw_point(x-dx, y+dy, width)
                self.draw_point(x-dy, y+dx, width)
                self.draw_point(x-dx, y-dy, width)
                self.draw_point(x-dy, y-dx, width)
                self.draw_point(x+dx, y-dy, width)
                self.draw_point(x+dy, y-dx, width)
            elif fill==FILL_FULL:
                for i in range(dx,dy):
                    self.draw_point(x+dx, y+i, width)
                    self.draw_point(x+i, y+dx, width)
                    self.draw_point(x-dx, y+i, width)
                    self.draw_point(x-i, y+dx, width)
                    self.draw_point(x-dx, y-i, width)
                    self.draw_point(x-i, y-dx, width)
                    self.draw_point(x+dx, y-i, width)
                    self.draw_point(x+i, y-dx, width)
            if d<0:
                d += 2*dx+3
            else:
                d += 2*(dx-dy)+5
                dy -= 1
            dx += 1
    def draw_str(self, x, y, text_str, font_size, font):
        str_list = list(font.getmask(text=text_str,mode="1"))
        str_size = font.getsize(text=text_str)
        font_act_h = int(len(str_list)/(str_size[0]))
        font_h = font_size
        top_add_row = int((font_h-font_act_h)/2)
        bot_add_row = int((font_h-font_act_h)) - top_add_row 
        for i in range(0,top_add_row*str_size[0]):
            str_list.insert(0,0)
        for i in range(0,bot_add_row*str_size[0]):
            str_list.append(0)
        for i in range(len(str_list)):
            if str_list[i]==0:
                str_list[i] = 1
            else:
                str_list[i] = 0
        n = 0
        for i in range(y-1,y+font_h-1):
            for j in range(x-1,x+str_size[0]-1):
                if str_list[n] == 0:
                    self.set_pixel(j,i)
                n += 1

File: arduino-mega/2inch13_epaper/epd_gui.cpp
#include <string.h>
#include "epd_2inch13.h"
#include <spi.h>
#include "epd_gui.h"
#include "fonts.h"

IMAGE Image;
void Image_init( uint16_t img_w, uint16_t img_h)
{
    memset(Image.img, 0xff, sizeof(Image.img));
    Image.w = img_w;
    Image.h = img_h;
}

//set the point(x,y) to black
//x:1~250 y:1~122
uint8_t Draw_point(uint16_t x, uint16_t y)
{
   uint8_t temp,bit_offset=0;
   uint16_t byte_offset=0;
   if(x>EPD_DIS_W || y>EPD_DIS_H)
   {
       printf("draw point out of the screen area!");
       return 0;
   }
   y += 6;//128-122=6
   byte_offset = (y-1)/8+(x-1)*16;
   bit_offset = (8-(y-1)%8)-1;
   temp = (uint8_t)(~(1<<(bit_offset)));
   Image.img[byte_offset] &= temp;
   return 1;
}

//clear the whole screen
void Gui_clear(void)
{
   memset(Image.img, 0xff, sizeof(Image.img));
}

//set pixel x:1~250 y:1~122
void Gui_set_pix(uint16_t x, uint16_t y)
{
    if(x>EPD_DIS_W || y>EPD_DIS_H)
    {
        printf("ERR:set_pix point out of screen area!\r\n");
        return ;
    }
    Draw_point( x, y);
}

void Gui_draw_point(uint16_t x, uint16_t y, uint16_t width)
{
    uint16_t xd, yd;
    if(x>EPD_DIS_W || y > EPD_DIS_H)
    {
        printf("ERR:Gui_draw_point out of screen area!\r\n");
        return ;
    }
    for(xd = 0;xd<width; xd++)
    {
        for(yd=0;yd<width; yd++)
        {
            if(x+xd>EPD_DIS_W || y+yd>EPD_DIS_H )
            {
               continue;    
            }
            Gui_set_pix(x+xd, y+yd);     
        }
    }
}
//Bresenham algorithm draw line
void Gui_draw_line(uint16_t x1, uint16_t y1, uint16_t x2, uint16_t y2, uint16_t width, LINE_STYLE type)
{
    uint16_t x = x1;
    uint16_t y = y1;
    int dx = x2-x1>=0?x2-x1:x1-x2;
    int dy = y2-y1>=0?y2-y1:y1-y2;
    int p = dx-dy;
    int e2;
    int xdir = x1>x2?-1:1;
    int ydir = y1>y2?-1:1;
    int dot_cnt = 0;
    for(;;)
    {
        dot_cnt++;
        if(type==DOTTED && dot_cnt%3==0)
        {
            Gui_draw_point(x, y, width);
        }
        else if(type==SOLID)
        {
            Gui_draw_point(x, y, width);
        }
        if(x==x2 && y==y2)
        {
            break;
        }
        e2 = 2*p;
        if(e2>= -dy)
        {
            p -= dy;
            x += xdir;
        }
        if(e2<=dx)
        {
            p += dx;
            y += ydir;
        }
    }
}

void Gui_draw_rectangle(uint16_t x1, uint16_t y1, uint16_t x2, uint16_t y2, uint16_t width, FILL fill)
{
    uint16_t i;
    if(x1>Image.w || y1 > Image.h || x2 > Image.w || y2 > Image.h)
    {
        printf("ERR:line point out of screen area!\r\n");
        return ;
    } 
    if(fill == FULL)
    {
        for(i=y1;i<y2;i++)
        {
            Gui_draw_line(x1, i, x2, i, width, SOLID);
        }
    }
    else if(fill==EMPTY)
    {
        Gui_draw_line(x1, y1, x2, y1, width, SOLID);
        Gui_draw_line(x1, y1, x1, y2, width, SOLID);
        Gui_draw_line(x1, y2, x2, y2, width, SOLID);
        Gui_draw_line(x2, y1, x2, y2, width, SOLID);
    }
}

//Bresenham  algorithm draw circle
void Gui_draw_circle(uint16_t x, uint16_t y, uint16_t r, uint16_t width, FILL fill)
{
    uint16_t dx=0,dy=r;
    int d = 1-r;
    uint16_t i;
    if(x>EPD_DIS_W || y > EPD_DIS_H)
    {
        printf("ERR:circle center point out of screen area!\r\n");
        return ;
    } 
    while(dy>dx)
    {
        if(fill==EMPTY)
        {
            Gui_draw_point(x+dx, y+dy, width);
            Gui_draw_point(x+dy, y+dx, width);
            Gui_draw_point(x-dx, y+dy, width);
            Gui_draw_point(x-dy, y+dx, width);
            Gui_draw_point(x-dx, y-dy, width);
            Gui_draw_point(x-dy, y-dx, width);
            Gui_draw_point(x+dx, y-dy, width);
            Gui_draw_point(x+dy, y-dx, width);
        }
        else if(fill==FULL)
        {
            for(i=dx;i<=dy;i++)
            {
                Gui_draw_point(x+dx, y+i, width);
                Gui_draw_point(x+i, y+dx, width);
                Gui_draw_point(x-dx, y+i, width);
                Gui_draw_point(x-i, y+dx, width);
                Gui_draw_point(x-dx, y-i, width);
                Gui_draw_point(x-i, y-dx, width);
                Gui_draw_point(x+dx, y-i, width);
                Gui_draw_point(x+i, y-dx, width);                
            }
        }
        if(d<0)
        {
            d += 2*dx+3;    
        }
        else
        {
            d += 2*(dx-dy)+5;
            dy--;
        }
        dx++;
    }
}

void Gui_draw_char(uint16_t x, uint16_t y, const char str_char, FONT* Font)
{
    uint16_t row, col_bit;
    uint32_t offset;
    uint16_t char_nbytes;
    const unsigned char *ptr;

    if (x > EPD_DIS_W || y > EPD_DIS_H) 
    {
        printf("ERR:draw char point out of screen area!\r\n");
        return;
    }
    //the number of bytes occupied by a character
    char_nbytes = Font->h * (Font->w / 8 + (Font->w % 8 ? 1 : 0));
    //the offset of the character
    offset = (str_char - ' ') * char_nbytes;
    ptr = &Font->table[offset];
    for (row = 0; row < Font->h; row ++) 
    {
        for (col_bit = 0; col_bit < Font->w; col_bit++ )
        {
            if (pgm_read_byte(ptr) & (0x80 >> (col_bit % 8)))
                Gui_set_pix(x+col_bit, y+row);            
            if (col_bit % 8 == 7) //next byte
                ptr++;
        }
        if(Font->w % 8 != 0)// next line
            ptr++;
    }
}

void Gui_draw_str(uint16_t x, uint16_t y, const char *str_char, FONT* Font)
{
    uint16_t dx = x,dy = y;
    while(*str_char != '\0')
    {
        if((dx+Font->w)>EPD_DIS_W)//wrap the line if the x direction has reached the boundary
        {
            dx = x;
            dy += Font->h;
        }
        if((dy+Font->h)>EPD_DIS_H)//wrap the line if the y direction has reached the boundary
        {
            dy = y;
            dx += Font->w;
        }
        Gui_draw_char(dx, dy, *str_char++, Font);
        dx += Font->w;
    }

}
